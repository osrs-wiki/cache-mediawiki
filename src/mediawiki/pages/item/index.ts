export { default as itemPageBuilder } from "./item";
export * from "./item.types";
