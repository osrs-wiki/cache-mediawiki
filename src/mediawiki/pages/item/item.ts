import {
  MediaWikiBreak,
  MediaWikiBuilder,
  MediaWikiFile,
  MediaWikiHeader,
  MediaWikiTemplate,
  MediaWikiText,
} from "@osrs-wiki/mediawiki-builder";

import { CombatStyles, InfoboxBonuses, InfoboxItem } from "../../templates";

import { Item, WearPos } from "@/utils/cache2";

export const itemPageBuilder = (item: Item) => {
  const infoboxItem = InfoboxItem(item);

  let infoboxBonuses = undefined;
  let combatStyles = undefined;
  if (item.wearpos1 >= 0) {
    infoboxBonuses = InfoboxBonuses(item);

    if (item.wearpos1 === WearPos.Weapon && item.category > -1) {
      combatStyles = CombatStyles(item);
    }
  }

  const builder = new MediaWikiBuilder();
  builder.addContents([
    new MediaWikiTemplate("New Content"),
    infoboxItem.build(),
    new MediaWikiFile(`${item.name} detail.png`, {
      horizontalAlignment: "left",
      resizing: { width: 130 },
    }),
    new MediaWikiBreak(),
    new MediaWikiText(item.name, { bold: true }),
  ]);

  if (infoboxBonuses) {
    builder.addContents([
      new MediaWikiBreak(),
      new MediaWikiBreak(),
      new MediaWikiHeader("Combat stats", 2),
      new MediaWikiBreak(),
      infoboxBonuses.build(),
      new MediaWikiBreak(),
    ]);

    if (combatStyles) {
      builder.addContents([combatStyles, new MediaWikiBreak()]);
    }
  }
  return builder;
};

export default itemPageBuilder;
