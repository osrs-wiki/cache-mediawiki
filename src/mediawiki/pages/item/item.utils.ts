import {
  MediaWikiBreak,
  MediaWikiContent,
  MediaWikiHeader,
  MediaWikiHTML,
  MediaWikiText,
} from "@osrs-wiki/mediawiki-builder";

import { Item } from "../../../utils/cache2";

/**
 * Generate a gallery of item stack variants.
 * @param item The item to generate the gallery for.
 * @returns An array of MediaWiki content for the gallery.
 */
export const generateItemStackGallery = (item: Item): MediaWikiContent[] => {
  const filteredStackVariants = item.stackVariantItems.filter((id) => id !== 0);
  const filteredStackQuantities = item.stackVariantQuantities.filter(
    (quantity) => quantity !== 0
  );
  if (filteredStackVariants.length === 0) {
    return [];
  }
  return [
    new MediaWikiBreak(),
    new MediaWikiBreak(),
    new MediaWikiHeader("Gallery", 2),
    new MediaWikiBreak(),
    new MediaWikiHTML(
      "gallery",
      [
        new MediaWikiText(`File:${item.name} 1 detail.png`),
        new MediaWikiText(`|1 ${item.name}`),
        new MediaWikiBreak(),
      ].concat(
        filteredStackVariants
          .map((_id, index) =>
            [
              new MediaWikiText(
                `File:${item.name}${
                  index < filteredStackQuantities.length - 1
                    ? ` ${filteredStackQuantities[index]}`
                    : ""
                } detail.png`
              ),
              new MediaWikiText(
                `|${filteredStackQuantities[index]} ${item.name}`
              ),
            ].concat(
              index < filteredStackVariants.length - 1
                ? [new MediaWikiBreak()]
                : []
            )
          )
          .flat()
      ),
      {
        captionalign: "center",
        widths: "120",
      }
    ),
  ];
};
