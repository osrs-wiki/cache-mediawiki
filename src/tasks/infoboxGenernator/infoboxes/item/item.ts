import {
  InfoboxTemplate,
  MediaWikiBreak,
  MediaWikiBuilder,
  MediaWikiDate,
  MediaWikiFile,
  MediaWikiHeader,
  MediaWikiTemplate,
} from "@osrs-wiki/mediawiki-builder";
import type { InfoboxItem } from "@osrs-wiki/mediawiki-builder";
import { mkdir, writeFile } from "fs/promises";

import {
  ATTACK_RANGE_PARAM,
  ATTACK_SPEED_PARAM,
  CRUSH_ATTACK_PARAM,
  CRUSH_DEFENCE_PARAM,
  MAGIC_ATTACK_PARAM,
  MAGIC_DAMAGE_PARAM,
  MAGIC_DEFENCE_PARAM,
  MELEE_STRENGTH_PARAM,
  PRAYER_BONUS_PARAM,
  RANGED_ATTACK_PARAM,
  RANGED_DEFENCE_PARAM,
  RANGED_STRENGTH_PARAM,
  SLASH_ATTACK_PARAM,
  SLASH_DEFENCE_PARAM,
  STAB_ATTACK_PARAM,
  STAB_DEFENCE_PARAM,
  formatBonus,
  getEquipmentSlot,
  getInventoryActions,
  getWeaponCategory,
} from "./item.utils";
import Context from "../../../../context";
import { CacheProvider, Item, WearPos } from "../../../../utils/cache2";

const itemInfoboxGenerator = async (
  cache: Promise<CacheProvider>,
  id: number
) => {
  try {
    const item = await Item.load(cache, id);

    buildItemInfobox(item);
  } catch (e) {
    console.error(`Error generating infobox for item ${id}: `, e);
  }
};

export const buildItemInfobox = async (item: Item, writeFiles = true) => {
  try {
    const infoboxItem = new InfoboxTemplate<InfoboxItem>("item", {
      name: item.name as string,
      image: new MediaWikiFile(`${item.name}.png`),
      release: Context.updateDate
        ? new MediaWikiDate(new Date(Context.updateDate))
        : undefined,
      update: Context.update,
      members: item.isMembers as boolean,
      quest: "No",
      exchange: item.isGrandExchangable,
      tradeable: item.isGrandExchangable,
      bankable: item.placeholderLinkedItem > 0,
      placeholder: item.placeholderLinkedItem > 1 ? true : undefined,
      equipable: item.wearpos1 > 0,
      stackable: item.isStackable,
      noteable: item.noteLinkedItem > 0,
      options: getInventoryActions(item),
      examine: item.examine,
      value: item.price,
      weight: (item.weight / 1000).toFixed(3),
      id: item.id.toString(),
    });

    let infoboxBonuses = undefined;
    let combatStyles = undefined;
    if (item.wearpos1 > 0) {
      infoboxBonuses = new InfoboxTemplate("bonuses", {
        astab: item.params.has(STAB_ATTACK_PARAM)
          ? formatBonus(item.params.get(STAB_ATTACK_PARAM))
          : "0",
        aslash: item.params.has(SLASH_ATTACK_PARAM)
          ? formatBonus(item.params.get(SLASH_ATTACK_PARAM))
          : "0",
        acrush: item.params.has(CRUSH_ATTACK_PARAM)
          ? formatBonus(item.params.get(CRUSH_ATTACK_PARAM))
          : "0",
        amagic: item.params.has(MAGIC_ATTACK_PARAM)
          ? formatBonus(item.params.get(MAGIC_ATTACK_PARAM))
          : "0",
        arange: item.params.has(RANGED_ATTACK_PARAM)
          ? formatBonus(item.params.get(RANGED_ATTACK_PARAM))
          : "0",
        dstab: item.params.has(STAB_DEFENCE_PARAM)
          ? formatBonus(item.params.get(STAB_DEFENCE_PARAM))
          : "0",
        dslash: item.params.has(SLASH_DEFENCE_PARAM)
          ? formatBonus(item.params.get(SLASH_DEFENCE_PARAM))
          : "0",
        dcrush: item.params.has(CRUSH_DEFENCE_PARAM)
          ? formatBonus(item.params.get(CRUSH_DEFENCE_PARAM))
          : "0",
        dmagic: item.params.has(MAGIC_DEFENCE_PARAM)
          ? formatBonus(item.params.get(MAGIC_DEFENCE_PARAM))
          : "0",
        drange: item.params.has(RANGED_DEFENCE_PARAM)
          ? formatBonus(item.params.get(RANGED_DEFENCE_PARAM))
          : "0",
        str: item.params.has(MELEE_STRENGTH_PARAM)
          ? formatBonus(item.params.get(MELEE_STRENGTH_PARAM))
          : "0",
        rstr: item.params.has(RANGED_STRENGTH_PARAM)
          ? formatBonus(item.params.get(RANGED_STRENGTH_PARAM))
          : "0",
        mdmg: item.params.has(MAGIC_DAMAGE_PARAM)
          ? `${parseInt(item.params.get(MAGIC_DAMAGE_PARAM).toString()) / 10}`
          : "0",
        prayer: item.params.has(PRAYER_BONUS_PARAM)
          ? formatBonus(item.params.get(PRAYER_BONUS_PARAM))
          : "0",
        slot: getEquipmentSlot(item),
        speed:
          item.wearpos1 === WearPos.Weapon &&
          item.params.has(ATTACK_SPEED_PARAM)
            ? item.params.get(ATTACK_SPEED_PARAM) ?? "0"
            : undefined,
        attackrange:
          item.wearpos1 === WearPos.Weapon &&
          item.params.has(ATTACK_RANGE_PARAM)
            ? item.params.get(ATTACK_RANGE_PARAM) ?? "0"
            : undefined,
        combatstyle: getWeaponCategory(item),
        image: new MediaWikiFile(`${item.name} equipped male.png`),
        altimage: new MediaWikiFile(`${item.name} equipped female.png`),
      });

      if (item.wearpos1 === WearPos.Weapon && item.category > -1) {
        combatStyles = new MediaWikiTemplate("CombatStyles");
        combatStyles.add(
          "speed",
          item.params.get(ATTACK_SPEED_PARAM)?.toString() ?? "0"
        );
        combatStyles.add(
          "attackrange",
          item.params.get(ATTACK_RANGE_PARAM)?.toString() ?? "0"
        );
        combatStyles.add("combatstyle", getWeaponCategory(item));
      }
    }

    const builder = new MediaWikiBuilder();
    builder.addContents([
      new MediaWikiTemplate("New Content"),
      infoboxItem.build(),
      new MediaWikiFile(`${item.name} detail.png`, {
        horizontalAlignment: "left",
        resizing: { width: 130 },
      }),
      new MediaWikiBreak(),
    ]);

    if (infoboxBonuses) {
      builder.addContents([
        new MediaWikiBreak(),
        new MediaWikiHeader("Combat stats", 2),
        new MediaWikiBreak(),
        infoboxBonuses.build(),
        new MediaWikiBreak(),
      ]);

      if (combatStyles) {
        builder.addContents([combatStyles, new MediaWikiBreak()]);
      }
    }

    if (writeFiles) {
      await mkdir("./out/infobox/item", { recursive: true });
      writeFile(`./out/infobox/item/${item.id}.txt`, builder.build());
    }
    return builder;
  } catch (e) {
    console.error("Error building item infobox: ", e);
  }
};

export default itemInfoboxGenerator;
