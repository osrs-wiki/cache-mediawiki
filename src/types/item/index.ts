export * from "./item";
export * from "./item.utils";
