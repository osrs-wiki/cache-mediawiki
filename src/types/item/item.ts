import { Item, ParamID, WearPos } from "../../utils/cache2";

export type WeaponSlot =
  | "head"
  | "weapon"
  | "body"
  | "legs"
  | "shield"
  | "cape"
  | "2h"
  | "hands"
  | "feet"
  | "neck"
  | "ammot"
  | "ring";

export type WeaponType =
  | "2h Sword"
  | "Axe"
  | "Banner"
  | "Bladed Staff"
  | "Blaster"
  | "Bludgeon"
  | "Blunt"
  | "Bow"
  | "Bulwark"
  | "Chinchompas"
  | "Claw"
  | "Crossbow"
  | "Gun"
  | "Multi-Style"
  | "Partisan"
  | "Pickaxe"
  | "Polearm"
  | "Polestaff"
  | "Powered Staff"
  | "Salamander"
  | "Scythe"
  | "Slash Sword"
  | "Spear"
  | "Spiked"
  | "Stab Sword"
  | "Staff"
  | "Thrown"
  | "Whip"
  | "Unarmed";
