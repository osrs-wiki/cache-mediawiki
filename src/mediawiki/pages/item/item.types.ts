import { MediaWikiFile } from "@osrs-wiki/mediawiki-builder";

import { WeaponSlot } from "../../../types/item/item";

export type InfoboxBonuses = {
  astab: string;
  aslash: string;
  acrush: string;
  amagic: string;
  arange: string;
  dstab: string;
  dslash: string;
  dcrush: string;
  dmagic: string;
  drange: string;
  str: string;
  rstr: string;
  mdmg: string;
  prayer: string;
  slot: WeaponSlot;
  speed?: number | "N/A";
  attackrange?: number | "No" | "staff";
  image?: MediaWikiFile;
  caption?: string;
  altimage?: MediaWikiFile;
  altcaption?: string;
};
