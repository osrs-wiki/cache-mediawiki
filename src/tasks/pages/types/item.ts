import Context from "../../../context";
import { renderItems } from "../../renders";
import { writePageToFile } from "../pages.utils";

import { itemPageBuilder } from "@/mediawiki/pages/item";
import { CacheProvider, Item } from "@/utils/cache2";

export const writeItemPageFromCache = async (
  cache: Promise<CacheProvider>,
  id: number
) => {
  try {
    const item = await Item.load(cache, id);

    if (item) {
      writeItemPage(item);
    }
  } catch (e) {
    console.error(`Error generating page for item ${id}: `, e);
  }
};

export const writeItemPage = async (item: Item) => {
  const builder = itemPageBuilder(item);
  writePageToFile(builder, "item", item.name, item.id.toString());

  if (Context.renders) {
    renderItems(item);
  }
};
