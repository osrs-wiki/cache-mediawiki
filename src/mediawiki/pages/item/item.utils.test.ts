import { MediaWikiBuilder } from "@osrs-wiki/mediawiki-builder";

import { generateItemStackGallery } from "./item.utils";
import { Item, ItemID } from "../../../utils/cache2";

describe("generateItemStackGallery", () => {
  const baseItem: Item = {
    name: "Test Item",
    stackVariantItems: [0, 2, 3],
    stackVariantQuantities: [3, 5, 10],
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any;

  it("generates gallery with stack variants", () => {
    const builder = new MediaWikiBuilder();
    builder.addContents(generateItemStackGallery(baseItem));
    expect(builder.build()).toMatchSnapshot();
  });

  it("handles empty stackVariantItems", () => {
    const item: Item = {
      ...baseItem,
      stackVariantItems: [],
      stackVariantQuantities: [],
    };
    const builder = new MediaWikiBuilder();
    builder.addContents(generateItemStackGallery(item));
    expect(builder.build()).toMatchSnapshot();
  });

  it("skips stackVariantItems with id 0", () => {
    const item: Item = {
      ...baseItem,
      stackVariantItems: [0 as ItemID, 0 as ItemID, 3 as ItemID],
      stackVariantQuantities: [2, 3, 4],
    };
    const builder = new MediaWikiBuilder();
    builder.addContents(generateItemStackGallery(item));
    expect(builder.build()).toMatchSnapshot();
  });
});
