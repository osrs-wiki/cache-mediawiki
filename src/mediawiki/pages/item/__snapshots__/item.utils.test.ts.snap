// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateItemStackGallery generates gallery with stack variants 1`] = `
"

==Gallery==
<gallery captionalign=\\"center\\", widths=\\"120\\">
File:Test Item 1 detail.png|1 Test Item
File:Test Item 3 detail.png|3 Test Item
File:Test Item 5 detail.png|5 Test Item
</gallery>
"
`;

exports[`generateItemStackGallery handles empty stackVariantItems 1`] = `""`;

exports[`generateItemStackGallery skips stackVariantItems with id 0 1`] = `
"

==Gallery==
<gallery captionalign=\\"center\\", widths=\\"120\\">
File:Test Item 1 detail.png|1 Test Item
File:Test Item 2 detail.png|2 Test Item
</gallery>
"
`;
